File: client/dtypes/cvalue.js
// CValue abstract class and related things.

var EventEmitter = require('events').EventEmitter;
var inherits = require('inherits');

////////////////////////////////////////////////////////////
// Events

// Base class for events.
function Event(isLocal) {
  this.isLocal = isLocal;
}

////////////////////////////////////////////////////////////
// CValue

inherits(CValue, EventEmitter);
function CValue() {
  EventEmitter.call(this);
  this.paused_ = false;
}

// Returns this value's dtype.
CValue.prototype.dtype = function() {
  throw new Error('not implemented');
};

// Applies the given encoded patch to this value.
// Similar to ApplyServerPatch in Go.
CValue.prototype.applyPatch = function(isLocal, patch) {
  throw new Error('not implemented');
};

// Returns true iff this value is paused, in which case local mutations are
// disallowed.
CValue.prototype.paused = function() {
  return this.paused_;
};

CValue.prototype.undo = function() {
  throw new Error('not implemented');
};

CValue.prototype.redo = function() {
  throw new Error('not implemented');
};

////////////////////////////////////////////////////////////
// Exports

module.exports = {
  CValue: CValue,
  dtypeCRegister: 'cregister',
  dtypeCString: 'cstring',
  dtypeDelete: 'delete',
  Event: Event
};

File: server/dtypes/cvalue/cvalue.go
// Package cvalue defines the CValue interface and related things.
package cvalue

import (
	"time"

	"github.com/asadovsky/cdb/server/common"
)

const (
	DTypeCRegister = "cregister"
	DTypeCString   = "cstring"
	DTypeDelete    = "delete"
)

// TODO: Switch to using []byte for encoded values, here and elsewhere.

// CValue is an interface for CDB values.
type CValue interface {
	// DType returns this value's dtype.
	DType() string

	// Encode encodes this value.
	Encode() (string, error)

	// ApplyServerPatch applies the given encoded patch to this value.
	ApplyServerPatch(patch string) error

	// ApplyClientPatch applies the given encoded patch to this value and returns
	// an encoded "server patch" suitable for persistent storage. The provided
	// patch may include client-only operations; the returned patch will never
	// contain such operations.
	ApplyClientPatch(agentId int, vec *common.VersionVector, t time.Time, patch string) (string, error)
}
